### PR TITLE
data filtering for rails backend

### PR DESCRIPTION
**Pull request by gw000**

### CODE REVIEW

File: backends/rails/lib/fetcher.rb [hunk 1]
```diff
@@ -9,12 +10,23 @@ def self.included(base)
         end
       end
 
-      kinds = Hash.new { |k| raise "Unsupported constraint kind: #{k}" }
-      kinds['Eq'] = kinds['In'] = ->(q, c) { q.where param[c] }
-      kinds['Neq'] = ->(q, c) { q.where.not param[c] }
+      it['Eq'] = it['In'] = ->(q, c) { q.where param[c] }
```
Comment (gkaemmer) [inline backends/rails/lib/fetcher.rb:13]: Feeling like this file might be sacrificing some readability for conciseness/performance. IMO the goal here is that someone can look at this file and use it to build their own fetchers, so it's probably worth lengthening variable names, adding some comments, etc.

I took a stab at restructuring it with some longer variable names and inlining the `instance_variable_set`. What do you think?
```ruby
module Fetcher
  def self.included(base) # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
    base.class_eval do
      to_activerecord_clause = lambda do |constraint|
        if constraint.field.nil?
          # When field is empty, we should match using ID
          { primary_key => constraint.value.send(primary_key) }
        else
          { c.field => c.value }
        end
      end

      @constraint_handlers = {
        'Eq': ->(query, constraint) { query.where to_activerecord_clause[constraint] },
        'In': ->(query, constraint) { query.where to_activerecord_clause[constraint] },
        'Neq': ->(query, constraint) { query.where.not to_activerecord_clause[constraint] }
      }
      @constraint_handlers.default_proc = proc { |k| "Unsupported constraint kind: #{k}" }
      @constraint_handlers.freeze

      def self.build_query(constraints)
        constraints.reduce(all) do |query, c|
          @constraint_handlers[constraint.kind][query, constraint]
        end
      end

      def self.exec_query(query)
        query.distinct.to_a
      end

      def self.combine_query(one, two)
        one.or(two)
      end
    end
  end
```

File: backends/rails/app/policy/authorization.polar [hunk 2]
```diff
@@ -42,29 +42,17 @@ resource(_type: Repo, "repo", actions, roles) if
...
-
-actor_has_role_for_resource(actor: User, role_name: String, resource: Repo) if
-    role in actor.repo_roles and
-    role_name = role.name and
-    resource.id = role.repo_id;
-
-actor_has_role_for_resource(org: Org, _role: String, repo: Repo) if
-    repo.org = org;
+actor_has_role_for_resource(_: User{repo_roles: roles}, name: String, repo: Repo) if
+    role in roles and role matches { name: name, repo: repo };
```
Comment (gkaemmer) [inline backends/rails/app/policy/authorization.polar:58]: Nice

File: backends/rails/app/controllers/repos_controller.rb [hunk 3]
```diff
@@ -2,9 +2,7 @@ class ReposController < ApplicationController
   def index
     org = Org.find(params[:org_id])
     authorize! "list_repos", org
-    # FIXME this isn't ideal but the current data filtering api is very rigid
-    # we'd like to be able to ask "what repos can this org show this user"
-    repos = OSO.get_allowed_resources(org, 'read', Repo)
+    repos = OSO.authorized_query(current_user, 'read', Repo).where(org: org)
```
Comment (gkaemmer) [inline backends/rails/app/controllers/repos_controller.rb:5]: I love this so much

**Review by gkaemmer — APPROVED** [review]
This looks good, my only feedback would be to add longer names and maybe some comments to `fetcher.rb` -- my guess is we'll be pointing a lot of people towards that file as a way to teach them the concepts

File: backends/rails/app/controllers/repos_controller.rb (hunk 3 above)
Comment (gj) [reply backends/rails/app/controllers/repos_controller.rb:5]: Why is the `.where(org: org)` clause necessary? Shouldn't that go in the Polar policy?

File: backends/rails/app/policy/authorization.polar [hunk 4]
```diff
@@ -56,12 +56,12 @@ allow(actor, action, resource) if
     role_allows(actor, action, resource);
 
 
-actor_has_role_for_resource(actor: User, role_name, resource: Org) if
+actor_has_role_for_resource(actor: User, role_name: String, resource: Org) if
     role in actor.org_roles and
     role_name = role.name and
-    resource = role.org;
+    resource.id = role.org_id;
```
[the remark this answers is not in the document]
Comment (gj) [reply backends/rails/app/policy/authorization.polar]: Did this happen?

File: backends/rails/app/controllers/repos_controller.rb (hunk 3 above)
Comment (gj) [reply backends/rails/app/controllers/repos_controller.rb:5]: Oh, I get it. The query is for all repos the user is allowed to read, but then we want to scope that down to only the current org.

File: backends/rails/app/policy/authorization.polar [hunk 5]
```diff
@@ -56,12 +56,15 @@ allow(actor, action, resource) if
...
 
-actor_has_role_for_resource(actor: User, role_name, resource: Repo) if
+actor_has_role_for_resource(actor: User, role_name: String, resource: Repo) if
     role in actor.repo_roles and
     role_name = role.name and
-    resource = role.repo;
+    resource.id = role.repo_id;
+
+actor_has_role_for_resource(org: Org, _role: String, repo: Repo) if
+    repo.org = org;
```
[the remark this answers is not in the document]
Comment (gj) [reply backends/rails/app/policy/authorization.polar]: @gw000 ping

File: backends/rails/lib/fetcher.rb (hunk 1 above)
Comment (gw000) [reply backends/rails/lib/fetcher.rb:13]: You're totally right, I'll change this 👍

File: backends/rails/app/policy/authorization.polar (hunk 5 above)
[the remark this answers is not in the document]
Comment (gw000) [reply backends/rails/app/policy/authorization.polar]: Oh hey sorry! With the new API this is out so it won't be a problem anymore, but the issue was that we needed to filter repos based on what org they belong to, and the only way to do that with data filtering was by expressing it as an allow rule. Now we can write something like `authorized_query(user, "read", Repo).where(org: org)`, which is much better.

File: backends/rails/app/policy/authorization.polar (hunk 4 above)
[the remark this answers is not in the document]
Comment (gj) [reply backends/rails/app/policy/authorization.polar]: cc: @gw000

**Comment by gj** [comment]
@gw000 looks like #15 is redundant after f4d7b94

**Pull request merged**